### ISSUE TITLE
hasMany through - filter/where/order not working

### ISSUE THREAD

**Issue by fabien** [strongloop/loopback-datasource-juggler#166]
```
var Space = app.loopback.getModel('Space');
var User = app.loopback.getModel('User');
var Account = app.loopback.getModel('Account');

Space.hasAndBelongsToMany('users', { through: Account });
Space.hasMany('accounts');

User.hasAndBelongsToMany(Space, { as: 'spaces', through: Account });
User.hasMany('accounts');
```

Example: `/api/spaces/client-a/users?filter[where][username]=toby` // result: []
Example: `/api/spaces/client-a/users?filter[order]=username` // result: unordered array

Regarding sorting, since models are fetched individually (byId vs. find), ordering is not applied either. This might very well be a limitation of how juggler handles relations at the moment (very flexible regarding mixed datasources, but not optimal performance wise).

I think the best approach would be to gather the related ids from the through-model first, then issue a second query on the actual target model, scoped to the previously retrieved ids - this is not only more efficient, but should enable the filtering capabilities as expected.

**Comment by superkhau** [comment]
Possibly related to https://github.com/strongloop/loopback-datasource-juggler/pull/476

**Comment by superkhau** [comment]
@raymondfeng Is this related to strongloop/loopback/issues/1076

**Comment by raymondfeng** [comment]
@superkhau I think so. Can you verify?

**Comment by superkhau** [comment]
>/api/spaces/client-a/users?filter[where][username]=toby // result: []

@fabien What is `client-a` in this example?

**Comment by fabien** [comment]
@superkhau it's just an `:id` value

**Comment by superkhau** [comment]
>Example: /api/spaces/client-a/users?filter[where][username]=toby // result: []
This issue has been resolved in strongloop/loopback#1076.

>Example: /api/spaces/client-a/users?filter[order]=username // result: unordered array
This issue is still unresolved due to the way we query related models. The current implementation performs a query on the join table and sorts each returned row individually instead of sorting all the resulting rows together as a whole.

Reassigned to @raymondfeng

**Comment by maelena87** [comment]
+1 @superkhau Example: /api/spaces/client-a/users?filter[order]=username // result: unordered array 
still not solved, right? still getting an unordered array.

**Comment by jannyHou** [comment]
I am working on this issue and create a repo to test the filters:
https://github.com/jannyHou/loopback-sandbox/tree/has-and-belongs-to-many

I tried 6 filters:
- fields
- include
- limit
- order
- skip
-where

`order` still doesn't work but other filters work properly.
I will fix `order` then. 

Details of my tests:

Three models in my app: `city`, 'space', 'customer'

1) 'city' hasMany 'space' & `space` belongsTo `city`
2) `space` hasAndBelongsToMany `customer`:
```
"relations": {
    "customers": {
      "type": "hasAndBelongsToMany",
      "model": "customer",
      "through": "spaceCustomer"
    }
  },
```

test cases:

- order(with problem): endpoint: `GET /spaces/{id}/customers`, filter: `{"order": "name"}`, result: unordered.

- fields: endpoint: `GET /spaces/{id}/customers`, filter: `{"fields": {"name": false}}`, result: without field `name`,
- include: endpoint: `GET /cities/{id}/spaces`, filter: `{"include": {"relation":"customers"}}`, result: all customers linked to space.
- limit: endpoint: `GET /spaces/{id}/customers`, filter: `{"limit": 2}`, result: only get 2 customers.
- skip: endpoint: `GET /spaces/{id}/customers`, filter: `{"skip": 2}`, result: skip first 2, only get 1 customers.
- where: endpoint: `GET /spaces/{id}/customers`, filter: `{"where": {"name": "zoe"}}`, result: only get customer zoe.

**Comment by jannyHou** [comment]
Update: 
`where` doesn't work when filter by id
- where:  endpoint: `GET /spaces/{id}/customers`, filter: `{"where": {"id": 1}}`, result: still get all customers.
I will fix it as well.

**Comment by ebarault** [comment]
it seems filter "fields" in hasManyThrough relation is still applied on through Model rather on target Model ?

**Comment by rpalacio86** [comment]
@ebarault same for "order" filter

**Comment by mplaza** [comment]
+1

**Comment by TeknoloGenie** [comment]
@jannyHou is this fix still being worked on? I can see that the limit works on filtering the relations but the sort still doesn't.

**Comment by dh376** [comment]
+1

**Comment by jannyHou** [comment]
@TeknoloGenie @ebarault @rpalacio86 @mplaza @dh376 
I still plan to fix this, hasn't implemented it yet since there is a workaround for this and I am busy on something else, sorry for that.

For example:
A relation `space` hasManyThrough `customer`
Query by `find` and apply `include` filter would work:
```js
// Works
// Result is ordered by `name`
Space.find({
      where: {id: space_id},
      include: {relation: 'customers', scope: {order: 'name'}}}, function(err, space){
      if (err) callback(err, null);
      callback(null, space);
 });
```

The signature doesn't work is queried by `findById` and directly apply filter order on related model:
```js
// Not work
     Space.findById(1, function(err, space){
       space.customers({
         order: 'name'
       }, function(err, result){
         if(err) callback(err, null);
         callback(null, result);
       });
     });
```

For a short term fix, could you use the first version instead? Fix the second one is not a simple job but I will work on it in the future for sure. Thanks for understanding:)

**Comment by PaddyMann** [comment]
+1 I have the same issue.

I have two models with a hasManyThrough relation: `Country` and `Partner`. I want to order the partners by the averageRating column in that model.

This works: /api/v2/partners?filter[order]=averageRating%20DESC
This does not: /api/v2/countries/GB/partners?filter[order]=averageRating%20DESC

Due to `ER_BAD_FIELD_ERROR: Unknown column 'averageRating' in 'order clause'`. SQL executed is "SELECT `id`,`countryId`,`partnerId` FROM `CountryPartner` WHERE `countryId`='GB' ORDER BY `averageRating` DESC"

I understand that there's a workaround if you jump into the code. But I'm expecting to use URLs like this in numerous places. Adding a new remote method in every location this is needed feels a bit crappy.

@jannyHou If I wanted to try and fix the underlying issue in loopback-datasource-juggler, where would I start? (it's not about to become my top priority, but would be useful to know if and when it does)

**Comment by jannyHou** [comment]
@PaddyMann thank you so much for trying to fix it, `loopback-datasource-juggler` contains complex logic in loopback, it's very easy to break other things even with a small change, I really appreciate your help but I'd better fix it myself so if I have any questions I can discuss with senior lb developers in time. Since many people are blocked by it, I label it as major here, will start to fix it next sprint.

**Comment by TeknoloGenie** [comment]
+1 @jannyHou thanks for the hard work! I've managed to find a workaround on the front-end to handle what i was needing for the time being, but it is moderately heavy, and won't work with large queries. So again, thanks for labeling this `major`!

**Comment by felipetoffolo1** [comment]
+1

**Comment by jreidal** [comment]
+1

**Comment by EvanWillms** [comment]
+1

**Comment by AndersonZacharyT** [comment]
+1

**Comment by ShaharHD** [comment]
+1

**Comment by thanickel** [comment]
+1

**Comment by thanickel** [comment]
@jannyHou If you can guide me where to look in the loopback-datasource-juggler code, I will fix this.

**Comment by RackyRose** [comment]
+1

**Comment by RLTmultimedia** [comment]
+1

**Comment by hoangtrongphuc** [comment]
Any update on this issues ?

**Comment by jannyHou** [comment]
@hoangtrongphuc I am working on a pr fixing it: https://github.com/strongloop/loopback-datasource-juggler/pull/1153

**Comment by jannyHou** [comment]
Hi @raymondfeng, the fix is published in 3.x and I am not sure should I backport it to 2.x?
I feel the filter would be good to have consistent behaviour in 2.x and 3.x, but the change itself is a bit breaking lol.
Any idea?

**Comment by raymondfeng** [comment]
@jannyHou What are the breaking changes? How much effort is needed to backport to 2.x?

**Comment by kristojorg** [comment]
I believe this may be still broken in 3.x . I have a TopUser extended from User which hasMany Role, but the response of `GET` on `/TopUser/:id/roles` does not show any roles.

**Comment by jannyHou** [comment]
@kristojorg could you fork https://github.com/strongloop/loopback-sandbox and provide me your code? Thanks. 
From your description, it's hard to tell is it broken by this fix or some other factors, for example, this issue is specific for relation `hasManyThrough` not `hasMany`, and some connector could also affect the behavior. 
With your sample app I can work with you to figure out how to get roles for a topuser.

**Comment by ilkerc** [comment]
+1

**Comment by sidewaiise** [comment]
Wow this is a big problem. {where} filter does not map to target model. I'd like to fix it in a fork.

Have looked at the libraries - wouldn't it make sense to have `RelationDefinition.hasManyThrough()` instead of trying to cram the through into `hasMany`? Seems it would be easier to manage from a library perspective.

Also please confirm - is the problem persisting in `hasAndBelongsToMany` relation? We did not use that because our datasource couldn't find the correct join table and/or relation labelings.

**Comment by jannyHou** [comment]
It's already released in 2.55.0 and also released long time ago in juggler@3.x, otherwise tests won't pass:
https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/relations.test.js#L726

Could anyone provide me a sandbox that reproduce the problem? see https://github.com/strongloop/loopback-datasource-juggler/issues/166#issuecomment-295861916

**Comment by ilkerc** [comment]
@jannyHou I tried. Couldn't reproduce. I'm using Mongo ObjectId's in every id field including reference id fields in relational tables. I think it's somehow related to this. If I catch anything related with mongo connector, shall I continue reporting here ?

**Comment by jannyHou** [comment]
Thank you @ilkerc, if it's a mongodb specific problem, then https://github.com/strongloop/loopback-connector-mongodb would be the best place to open issue and discuss.

**Comment by mbenedettini** [comment]
@jannyHou as far as I can see, those tests you mentioned only test the case when filter is an id. If I change the where condition to be, for example, name or any other attribute, then the query stops working.

**Comment by jannyHou** [comment]
@mbenedettini Thank you I think it's caused by same reason as issue https://github.com/strongloop/loopback/issues/3487 reported. 
And there is a PR fix opened for it https://github.com/strongloop/loopback-datasource-juggler/pull/1406

I am going to close this issue since the original bug is fixed in [juggler@3.1.0](https://github.com/strongloop/loopback-datasource-juggler/blob/master/CHANGES.md#2016-12-05-version-310), since the failures you observe is caused by another bug.
The conversation happens here is quite misleading :( I find some similar issues opened in loopback and loopback-datasource-juggler, will post another story to track the new bug&fix later.
Thanks for understanding.

**Comment by jannyHou** [comment]
Let's keep discussion in issue https://github.com/strongloop/loopback/issues/3487, which should reflect the filter issue you run into recently.
